astropy/utils/metadata.py

```diff
@@ -23,26 +24,129 @@ class MergeConflictWarning(AstropyWarning):
+
+    # For string-type arrays need to explicitly fill in non-zero
+    # values or the final arr_common = .. step is unpredictable.
+    for arr in arrs:
+        if arr.dtype.kind in ('S', 'U'):
+            arr[0] = '0' * arr.itemsize
+
+    arr_common = np.array([arr[0] for arr in arrs])
+    return arr_common.dtype.str
+
+
+class MergeStrategyMeta(type):
+    """
+    Metaclass that registers MergeStrategy subclasses into the
+    MERGE_STRATEGY_CLASSES registry.
     """
-    Concatenate the ``left`` and ``right`` objects
 
-    Currently only lists and tuples are allowed, but this could change.
+    def __new__(mcls, name, bases, members):
+        cls = super(MergeStrategyMeta, mcls).__new__(mcls, name, bases, members)
+
+        # Register merging class (except for base MergeStrategy class)
+        if 'types' in members:
+            types = members['types']
+            if isinstance(types, tuple):
+                types = [types]
+            for left, right in reversed(types):
+                MERGE_STRATEGY_CLASSES.insert(0, (left, right, cls))
+
+        return cls
+
+@six.add_metaclass(MergeStrategyMeta)
+class MergeStrategy(object):
+    # Set to False to disable applying this merge strategy
+    enabled = True
+
+    @classmethod
+    def merge_with_catch(cls, left, right):
+        """
+        Call the ``merge()`` method, but catch exceptions and re-raise
+
+        Parameters
+        ----------
+        left: object
+            Left object to be merged
+        right: object
+            Right object to be merged
+
+        Returns
+        -------
+        out: object
+            Merged left and right object
+        """
+        try:
+            return cls.merge(left, right)
+        except Exception as err:
+            raise MergeConflictError(err)
+
+
+class MergePlus(MergeStrategy):
+    """
+    Merge ``left`` and ``right`` objects using the plus operator.
     """
-    if type(left) != type(right):
-        raise MergeConflictError()
-    if any(isinstance(left, cls) for cls in (tuple, list)):
+    types = [(list, list),
+             (tuple, tuple)]
+
+    @classmethod
+    def merge(cls, left, right):
         return left + right
-    else:
-        raise MergeConflictError()
+
+
+class MergeNpConcatenate(MergeStrategy):
+    """
+    Merge ``left`` and ``right`` objects using np.concatenate.
+
+    This will upcast a list or tuple to np.ndarray and the output is
+    always ndarray.
+    """
+    types = [(np.ndarray, np.ndarray),
+             (np.ndarray, (list, tuple)),
+             ((list, tuple), np.ndarray)]
+
+    @classmethod
+    def merge(cls, left, right):
+        left, right = np.asanyarray(left), np.asanyarray(right)
+        common_dtype([left, right])  # Ensure left and right have compatible dtype
+        return np.concatenate([left, right])
+
+
+def take_left(left, right):
+    return left
+
+
+def take_right(left, right):
+    return left
```

Comment:
I actually think those functions are orphaned from some (old) previous ideas of where metadata merging was going.  The fact that there is an obvious bug makes me pretty confident that they are not being used, so I'm just going to remove them.